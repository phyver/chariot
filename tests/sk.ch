val k x y = x

val s x y z = x z (y z)

val i = s k k

:show functions


(* AMEN combinators *)

val a i j f x = i f (j f x)

val m i j f = i (j f)

val e i j = j i

val n f x = x

:show functions

val one = e n n

val two = a one one

val three = a one two

val four = m two two

val nine = e three two


data nat where Zero : nat | Succ : nat -> nat

:reduce one Succ Zero

:reduce two Succ Zero

:reduce three Succ Zero

:reduce four Succ Zero

:reduce nine Succ Zero


