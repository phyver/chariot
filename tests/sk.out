>> Typing for k successful
>> the definition for k is complete
>> the definition k is provably correct

>> Typing for s successful
>> the definition for s is complete
>> the definition s is provably correct

>> Typing for i successful
>> the definition for i is complete
>> the definition i is provably correct

>> functions in environment:
val
   k : 'a → 'b → 'a
    | k x y = x

val
   s : ('a → 'b → 'c) → ('a → 'b) → 'a → 'c
    | s x y z = x z (y z)

val
   i : 'a → 'a
    | i = s k k

>> Typing for a successful
>> the definition for a is complete
>> the definition a is provably correct

>> Typing for m successful
>> the definition for m is complete
>> the definition m is provably correct

>> Typing for e successful
>> the definition for e is complete
>> the definition e is provably correct

>> Typing for n successful
>> the definition for n is complete
>> the definition n is provably correct

>> functions in environment:
val
   k : 'a → 'b → 'a
    | k x y = x

val
   s : ('a → 'b → 'c) → ('a → 'b) → 'a → 'c
    | s x y z = x z (y z)

val
   i : 'a → 'a
    | i = s k k

val
   a : ('a → 'b → 'c) → ('a → 'd → 'b) → 'a → 'd → 'c
    | a i j f x = i f (j f x)

val
   m : ('a → 'b) → ('c → 'a) → 'c → 'b
    | m i j f = i (j f)

val
   e : 'a → ('a → 'b) → 'b
    | e i j = j i

val
   n : 'a → 'b → 'b
    | n f x = x

>> Typing for one successful
>> the definition for one is complete
>> the definition one is provably correct

>> Typing for two successful
>> the definition for two is complete
>> the definition two is provably correct

>> Typing for three successful
>> the definition for three is complete
>> the definition three is provably correct

>> Typing for four successful
>> the definition for four is complete
>> the definition four is provably correct

>> Typing for nine successful
>> the definition for nine is complete
>> the definition nine is provably correct

>> inductive type nat was succesfully defined
>> term: one Succ 0
>> result: 1
>> of type: nat

>> term: two Succ 0
>> result: 2
>> of type: nat

>> term: three Succ 0
>> result: 3
>> of type: nat

>> term: four Succ 0
>> result: 4
>> of type: nat

>> term: nine Succ 0
>> result: 9
>> of type: nat

